describe how the tools work and how they fit together

smason opened the pull request:
target audience is mostly developers / those wanting to understand the code base

probably want some more detailed user facing documentation somewhere as well

smason committed:
initial attempt at an ARCHITECTURE doc

smason committed:
improvement writing

johnnonweiler commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,183 @@
+
+# Terminology
+
+ * A **solid**s is a 3D volume described by a set of faces, edges, and
```
```suggestion
 * A **solid** is a 3D volume described by a set of faces, edges, and
```

johnnonweiler commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,183 @@
+
+ * A **solid**s is a 3D volume described by a set of faces, edges, and
+   verticies.
+ * Solids **intersect** when some point exists that is within either
+   solids or on the surface of either solid. It's useful to
+   distinguish between these cases:
```
```suggestion
 * Solids **intersect** when some point exists that is within (or on the surface of) both solids. It's useful to
   distinguish between these cases:
```

johnnonweiler commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,183 @@
+ * Solids are **distinct** when they do not intersect in any way. For
+   example, their bounding boxes do not overlap. Note that this can
+   get awkward to determine, e.g. a cylinder could have a torus around
+   it whose "hole" is larger diameter than the cylinder.
```
```suggestion
   it whose "hole" has a larger diameter than the cylinder.
```

johnnonweiler commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,183 @@
+reference the parts consistently.
+
+CAD models tend to be hierarchical (it makes it easier to work on them
+like this) but this makes referring to parts awkward. The way I've
+chosen is to "linearise" the hierarchy of solids into a list, and
+maintain this linear order across related files and operations.
```
A key point here is that you're not creating any new solids (e.g. with solid1, solid2, and intersection12), but are adding the intersection to one of the original solids.

johnnonweiler replied:
Another question is whether you can get the solids out of the STEP file in a deterministic order?

smason replied:
getting files out of a STEP file in deterministic order shouldn't matter, that's just done once when converting to BREP.  it's whether running these different utilities iterate over the BREP solids they get them back in the same order

smason replied:
> A key point here

this was what the following paragraph (i.e. "For example, when checking") is giving examples of?

johnnonweiler commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,183 @@
+# Overlap checking
+
+Once the solids have been linearised, it's a simple matter of
+performing $N(N-1)$ checks to determine which solids intersect with
```
What does the $N(N-1)$ syntax represent?  (It looks like some kind of Markdown formatting, but GitHub doesn't seem to interpret it as that.)

smason replied:
oops, forgot about that.  Jupyter recognizes it in their notebooks as a latex style equation, forgot that github doesn't support it!

johnnonweiler commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,183 @@
+# Model clean-up
+
+For a model to be used in (e.g.) neutronics modelling, we need to
+remove any overlaps between solids (`imprint_solids`) and merge
+(`merge_solids`) vertexes/edges/faces where appropriate.
```
```suggestion
remove any overlaps between solids (`imprint_solids`) and merge
vertexes/edges/faces (`merge_solids`) where appropriate.
```

johnnonweiler commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,183 @@
+shared faces. For example, two adjacent cubes should share adjacent
+faces to indicate they touch each other. Otherwise the model might
+interpret this as needing a particle to transition from solid, into
+air, then back into a solid, or maybe that the two solids can slide
+past each other.
```
I don't understand the bit about two solids sliding past each other.

johnnonweiler reviewed (commented):
This seems very clear and helpful to me.

smason committed:
improvements suggested by @johnnonweiler

smason committed:
remove reference to solids sliding

john think's it's confusing considering everything else is about
neutronics pipelines
